libraries/freertos_plus/standard/freertos_plus_tcp/source/portable/NetworkInterface/Common/phyHandling.c

```diff
@@ -0,0 +1,714 @@
+
+/* Discover all PHY's connected by polling 32 indexes ( zero-based ) */
+BaseType_t xPhyDiscover( EthernetPhy_t *pxPhyObject )
+{
+BaseType_t xPhyAddress;
+
+	pxPhyObject->xPortCount = 0;
+
+	for( xPhyAddress = phyMIN_PHY_ADDRESS; xPhyAddress <= phyMAX_PHY_ADDRESS; xPhyAddress++ )
+	{
+	uint32_t ulLowerID;
+
+		pxPhyObject->fnPhyRead( xPhyAddress, phyREG_03_PHYSID2, &ulLowerID );
+		/* A valid PHY id can not be all zeros or all ones. */
+		if( ( ulLowerID != ( uint16_t )~0u )  && ( ulLowerID != ( uint16_t )0u ) )
+		{
+		uint32_t ulUpperID;
+		uint32_t ulPhyID;
+
+			pxPhyObject->fnPhyRead( xPhyAddress, phyREG_02_PHYSID1, &ulUpperID );
+			ulPhyID = ( ( ( uint32_t ) ulUpperID ) << 16 ) | ( ulLowerID & 0xFFF0 );
+
+			pxPhyObject->ucPhyIndexes[ pxPhyObject->xPortCount ] = xPhyAddress;
+			pxPhyObject->ulPhyIDs[ pxPhyObject->xPortCount ] = ulPhyID;
+
+			pxPhyObject->xPortCount++;
+
+			/* See if there is more storage space. */
+			if( pxPhyObject->xPortCount == ipconfigPHY_MAX_PORTS )
+			{
+				break;
+			}
+		}
+	}
+	if( pxPhyObject->xPortCount > 0 )
+	{
+		FreeRTOS_printf( ( "PHY ID %lX\n", pxPhyObject->ulPhyIDs[ 0 ] ) );
+	}
+
+	return pxPhyObject->xPortCount;
+}
+/*-----------------------------------------------------------*/
+
+/* Send a reset commando to a set of PHY-ports. */
+static uint32_t xPhyReset( EthernetPhy_t *pxPhyObject, uint32_t ulPhyMask )
+{
+uint32_t ulDoneMask, ulConfig;
+TickType_t xRemainingTime;
+TimeOut_t xTimer;
+BaseType_t xPhyIndex;
+
+	/* A bit-mask of PHY ports that are ready. */
+	ulDoneMask = 0ul;
+
+	/* Set the RESET bits high. */
+	for( xPhyIndex = 0; xPhyIndex < pxPhyObject->xPortCount; xPhyIndex++ )
+	{
+	BaseType_t xPhyAddress = pxPhyObject->ucPhyIndexes[ xPhyIndex ];
+
+		/* Read Control register. */
+		pxPhyObject->fnPhyRead( xPhyAddress, phyREG_00_BMCR, &ulConfig );
+		pxPhyObject->fnPhyWrite( xPhyAddress, phyREG_00_BMCR, ulConfig | phyBMCR_RESET );
+	}
+
+	xRemainingTime = ( TickType_t ) pdMS_TO_TICKS( phyPHY_MAX_RESET_TIME_MS );
+	vTaskSetTimeOutState( &xTimer );
+
+	/* The reset should last less than a second. */
+	for( ;; )
+	{
+		for( xPhyIndex = 0; xPhyIndex < pxPhyObject->xPortCount; xPhyIndex++ )
+		{
+		BaseType_t xPhyAddress = pxPhyObject->ucPhyIndexes[ xPhyIndex ];
+
+			pxPhyObject->fnPhyRead( xPhyAddress, phyREG_00_BMCR, &ulConfig );
+			if( ( ulConfig & phyBMCR_RESET ) == 0 )
+			{
+				FreeRTOS_printf( ( "xPhyReset: phyBMCR_RESET %d ready\n", (int)xPhyIndex ) );
+				ulDoneMask |= ( 1ul << xPhyIndex );
+			}
+		}
+		if( ulDoneMask == ulPhyMask )
+		{
+			break;
+		}
+		if( xTaskCheckForTimeOut( &xTimer, &xRemainingTime ) != pdFALSE )
+		{
+			FreeRTOS_printf( ( "xPhyReset: phyBMCR_RESET timed out ( done 0x%02lX )\n", ulDoneMask ) );
+			break;
+		}
+		/* Block for a while */
+		vTaskDelay( pdMS_TO_TICKS( phySHORT_DELAY_MS ) );
+	}
+
+	/* Clear the reset bits. */
+	for( xPhyIndex = 0; xPhyIndex < pxPhyObject->xPortCount; xPhyIndex++ )
```

Comment:
Is this needed only in case of failure or in normal execution as well? If it is needed only for failure, can we avoid executing it for success?

Reply:
@aggarg, thanks for your comments.

> phyHandling.h 
> *pulValue as opposed to *pusValue.

Done.

> phyHandling.h 
> ulValue as opposed to usValue.

Done.

> phyHandling.h

commando => command in 2 occurrences

Done.

> `/* ID's of supported PHY's : */`
> `#define PHY_ID_LAN8742A		0x0007c130`
> These IDs probably correspond to the supported hardware.
> May be helpful to mention those hardware in comments.

Not sure what you mean here?
In this example, the hardware is 'LAN8742A', a PHY driver compatible with any CPU that supports RMII.


~~~c
static __inline uint32_t xPhyGetMask( EthernetPhy_t *pxPhyObject )
{
	return ( ( ( uint32_t ) 1u ) << pxPhyObject-> xPortCount ) - 1;
}
~~~

From now on it is a `#define`.

> phyHandling.c

LinkSStatus => LinkStatus in 2 occurrences

Done.

> phyADVERTISE_ALL does not include phyADVERTISE_CSMA?

I have included it now.

> `/* Clear the reset bits. */`
> Is this needed only in case of failure or in normal
> execution as well? If it is needed only for failure,
> can we avoid executing it for success?

Good catch: it is useless to clear a flag when the reset operation succeeded.

I retested the changes in a STM32F4 board that has a LAN8720 transceiver.

Thanks